src/Controller/SigningController.php

```diff
@@ -0,0 +1,108 @@
+<?php
+
+namespace Drupal\os2forms_dig_sig_server\Controller;
+
+use Drupal\Core\Controller\ControllerBase;
+use Drupal\os2forms_dig_sig_server\Service\SigningService;
+use Exception;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+use Symfony\Component\HttpFoundation\JsonResponse;
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpKernel\Exception\BadRequestHttpException;
+
+class SigningController extends ControllerBase {
+
+  public function __construct(
+    private readonly SigningService $signingService,
+    private readonly RequestStack $requestStack,
+  ) {
+  }
+
+  public static function create(ContainerInterface $container): static {
+    return new static(
+      $container->get('os2forms_dig_sig_server.signing_service'),
+      $container->get('request_stack'),
+    );
+  }
+
+  /**
+   * Handles signing-related actions based on the 'action' parameter in the request.
+   *
+   * Supports various operations:
+   * - 'getcid': Retrieves a correlation ID.
+   * - 'sign': Initiates a signing process for the specified URI, forward URL, and hash.
+   * - 'cancel' or 'result': Processes the result or cancels an operation using information from the specified file.
+   * - 'download': Downloads a file, optionally retaining it based on the 'leave' parameter.
+   *
+   * Throws an exception for unexpected or invalid actions or if required parameters are missing.
+   *
+   * @return Response
+   *    Returns a JSON response containing the result of the specified action.
+   *    Or TrustedRedirectResponse.
+   */
+  public function signingCallback() {
+    $request = $this->requestStack->getCurrentRequest();
+    $action = $request->query->get('action');
+
+    try {
+      switch ($action) {
+        case 'getcid':
+          $json = ['cid' => $this->signingService->getCorrelationId()];
+          return new JsonResponse($json);
+
+        case 'sign':
+          $uri = $request->query->get('uri');
+          $forward_url = $request->query->get('forward_url');
```

Comment:
Consider camel casing this variable name:

```suggestion
          $forwardUrl = $request->query->get('forward_url');
```